drivers/i2c/i2c_nrfx_twim.c

```diff
@@ -52,18 +61,42 @@ static int i2c_nrfx_twim_transfer(struct device *dev, struct i2c_msg *msgs,
 			break;
 		}
 
-		nrfx_twim_xfer_desc_t cur_xfer = {
-			.p_primary_buf  = msgs[i].buf,
-			.primary_length = msgs[i].len,
-			.address	= addr,
-			.type		= (msgs[i].flags & I2C_MSG_READ) ?
-					  NRFX_TWIM_XFER_RX : NRFX_TWIM_XFER_TX
-		};
+		if ((concat_len != 0)
+		    || ((i + 1 < num_msgs)
+			&& !(msgs[i].flags & I2C_MSG_STOP)
+			&& !(msgs[i + 1].flags & I2C_MSG_RESTART))) {
+			if (msgs[i].len > concat_buf_size) {
+				LOG_ERR("Concatenation buffer is too small");
+				return -ENOSPC;
+			}
+			if (!(msgs[i].flags & I2C_MSG_READ)) {
+				memcpy(concat_buf + concat_len,
+				       msgs[i].buf,
+				       msgs[i].len);
+			}
+			concat_len += msgs[i].len;
+		}
+
+		if ((msgs[i].flags & I2C_MSG_STOP)
+		    || !(i + 1 < num_msgs)
+		    || (msgs[i + 1].flags & I2C_MSG_RESTART)) {
```

Comment:
I think it would be good to introduce a flag like:
```
		bool last_or_non_concatenable =
			((msgs[i].flags & I2C_MSG_STOP)
			 || !(i + 1 < num_msgs)
			 || (msgs[i + 1].flags & I2C_MSG_RESTART)
			 || ((msgs[i + 1].flags & I2C_MSG_READ)
			     != (msgs[i].flags & I2C_MSG_READ)));
```
(in my previous comment I forgot about the last part - checking if the next message is a transfer in the same direction; I think it should be checked here, even though the change of direction should come together with an explicit STOP or RESTART flag).
And this flag could be used here and in the check above:
```
		if ((concat_len != 0) || !last_or_non_concatenable)
```
This way the code will be also a bit easier to comprehend.